File: sliding-window/character-replacement.py
from collections import defaultdict


# O(1) since finite chars
def maxFreq(charFreqs: dict[str, int]):
    maxFreq = 0
    for c in charFreqs:
        if charFreqs[c] > maxFreq:
            maxFreq = charFreqs[c]

    return maxFreq


class Solution:

    def characterReplacement(self, s: str, k: int) -> int:
        n = len(s)
        left = 0
        right = 0
        charFreqs = defaultdict(int)
        maxLen = 0

        while right < n:
            charFreqs[s[right]] += 1
            right += 1

            while right-left-maxFreq(charFreqs) > k:
                charFreqs[s[left]] -= 1
                left += 1

            print(right, left)
            maxLen = max(maxLen, right-left)

        return maxLen

File: sliding-window/character-replacement.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

int replacementsNeeded(unordered_map<char, int> &umap)
{
    int charCount = 0;
    int maxCount = 0;
    for (auto &pair : umap)
    {
        charCount += pair.second;
        maxCount = max(maxCount, pair.second);
    }
    return charCount - maxCount;
}

class Solution
{
public:
    int characterReplacement(string s, int k)
    {
        unordered_map<char, int> freqs;
        int n = s.size();
        int left = 0;
        int right = 0;
        int maxL = 0;
        while (right < n)
        {
            if (freqs.find(s[right]) == freqs.end())
            {
                freqs[s[right]] = 0;
            }
            freqs[s[right]]++;
            right++;

            while (replacementsNeeded(freqs) > k)
            {
                freqs[s[left]]--;
                left++;
            }
            maxL = max(maxL, right - left);
        }
        return maxL;
    }
};
